src/devices_models/devices/common/add_constraint_dual.jl

```diff
@@ -44,15 +44,40 @@ function add_constraint_dual!(
     sys::PSY.System,
     model::ServiceModel{T, D},
 ) where {T <: PSY.Service, D <: AbstractServiceFormulation}
+    if !isempty(get_duals(model))
+        service = PSY.get_component(T, sys, model.service_name)
+        for constraint_type in get_duals(model)
+            assign_dual_variable!(container, constraint_type, service, D())
+        end
+    end
+    return
+end
+
+function assign_dual_variable!(
+    container::OptimizationContainer,
+    constraint_type::Type{<:ConstraintType},
+    service::D,
+    ::AbstractServiceFormulation,
+) where {D <: PSY.Service}
+    time_steps = get_time_steps(container)
+    service_name = PSY.get_name(service)
+    add_dual_container!(
+        container,
+        constraint_type,
+        D,
+        [service_name],
+        time_steps;
+        meta=service_name,
+    )
     return
 end
 
 function assign_dual_variable!(
     container::OptimizationContainer,
     constraint_type::Type{<:ConstraintType},
     devices::U,
-    formulation,
-) where {U <: Union{Vector{D}, IS.FlattenIteratorWrapper{D}}} where {D <: PSY.Component}
+    ::AbstractDeviceFormulation,
+) where {U <: Union{Vector{D}, IS.FlattenIteratorWrapper{D}}} where {D <: PSY.Device}
```

Comment:
Should this still be `Component`? The error I got seems to indicate it's trying to apply this function to `buses`?

Reply:
Not all components have device formulations. In my last commit I addressed this.